plugins/coordinator/milestones.go

```diff
@@ -0,0 +1,342 @@
+func doPow(tx *transaction.Transaction, mwm int) error {
+
+	tx.AttachmentTimestamp = time.Now().UnixNano() / int64(time.Millisecond)
+	tx.AttachmentTimestampLowerBound = consts.LowerBoundAttachmentTimestamp
+	tx.AttachmentTimestampUpperBound = consts.UpperBoundAttachmentTimestamp
+
+	trytes, err := transaction.TransactionToTrytes(tx)
+	if err != nil {
+		return err
+	}
+
+	nonce, err := powFunc(trytes, mwm)
+	if err != nil {
+		return err
+	}
+
+	tx.Nonce = nonce
+
+	// set new transaction hash
+	tx.Hash = transactionHash(tx)
+
+	return nil
+}
+
+// transactionHash makes a transaction hash from the given transaction.
+func transactionHash(t *transaction.Transaction) trinary.Hash {
+	trits, _ := transaction.TransactionToTrits(t)
+	hashTrits := batchhasher.CURLP81.Hash(trits)
+	return trinary.MustTritsToTrytes(hashTrits)
+}
+
+// finalizeInsecure sets the bundle hash for all transactions in the bundle
+// we do not care about the M-Bug since we use a fixed version of the ISS
+func finalizeInsecure(bundle Bundle) (Bundle, error) {
+
+	k := kerl.NewKerl()
+
+	for _, tx := range bundle {
+		txTrits, err := transaction.TransactionToTrits(tx)
+		if err != nil {
+			return nil, err
+		}
+
+		k.Absorb(txTrits[consts.AddressTrinaryOffset:consts.BundleTrinaryOffset]) // Address + Value + ObsoleteTag + Timestamp + CurrentIndex + LastIndex
+	}
+
+	bundleHashTrits, err := k.Squeeze(consts.HashTrinarySize)
+	if err != nil {
+		return nil, err
+	}
+	bundleHash := trinary.MustTritsToTrytes(bundleHashTrits)
+
+	// set the computed bundle hash on each tx in the bundle
+	for _, tx := range bundle {
+		tx.Bundle = bundleHash
+	}
+
+	return bundle, nil
+}
+
+// chainTransactionsFillSignatures fills the signature message fragments with the signature and sets the trunk to chain the txs in a bundle
+func chainTransactionsFillSignatures(b Bundle, signature trinary.Trytes, mwm int) error {
+	// to chain transactions we start from the LastIndex and move towards index 0.
+	prev := b[len(b)-1].Hash
+
+	// we have to skip the siblingsTx, because it is already complete
+	for i := len(b) - 2; i >= 0; i-- {
+		tx := b[i]
+
+		// copy signature fragment
+		tx.SignatureMessageFragment = signature[tx.CurrentIndex*consts.SignatureMessageFragmentSizeInTrytes : (tx.CurrentIndex+1)*consts.SignatureMessageFragmentSizeInTrytes]
+
+		// chain bundle
+		tx.TrunkTransaction = prev
+
+		// perform PoW
+		if err := doPow(tx, mwm); err != nil {
+			return err
+		}
+
+		prev = tx.Hash
+	}
+	return nil
+}
+
+// issueCheckpoint sends a secret checkpoint transaction to the network
+// we do that to prevent parasite chain attacks
+// only honest tipselection will reference our checkpoints, so the milestone will reference honest tips
+func issueCheckpoint(lastCheckpointHash *trinary.Hash) (trinary.Hash, error) {
+
+	tips, _, err := tipselection.SelectTips(0, lastCheckpointHash)
+	if err != nil {
+		return "", err
+	}
+
+	tx := &transaction.Transaction{}
```

Comment:
shouldn't you use something more random than just null trytes everywhere? you are always creating the same tx, just changing the Timestamp. I think this is something that could be attacked due to the fact that you can estimate when the coordinator will generate the next milestone and its checkpoints, so you can reattach the checkpoints on your parasite chain since they are easy to identify